extensions/libevent.c

```diff
@@ -0,0 +1,318 @@
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 3 of
+ * the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA
+ */
+
+
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <event2/event.h>
+#include <libguile.h>
+
+struct event_data
+{
+  int fd;
+  short event;
+};
+
+struct wait_data
+{
+  int rv;
+  struct event_data *events;
+  int maxevents;
+};
+
+struct loop_data
+{
+  struct event_base *base;
+  int64_t timeout;
+};
+
+struct libevt_data
+{
+  struct event_base *base;
+};
+
+
+static void
+free_libevt (void *ptr)
+{
+  struct libevt_data *libevt = ptr;
+  event_base_free (libevt->base);
+  scm_gc_free(libevt, sizeof (struct libevt_data), "libevt_data");
+}
+
+static void
+free_wait_data (void *ptr)
+{
+  struct wait_data *data = ptr;
+  scm_gc_free(data, sizeof (struct wait_data), "wait_data");
+}
+
+static void
+cb_func (evutil_socket_t fd, short what, void *arg)
+{
+  struct wait_data* data = arg;
+  int rv = data->rv;
+
+  // In theory, we should never exceed the maximum number of events. This is
+  // because every time we add a new FD to monitor we check if we have enough
+  // room in our vector and if not we resize it. So, if this condition is true
+  // is because those assumptions are wrong and we might be doing something
+  // funky.
+  if (rv >= data->maxevents)
+    {
+      scm_puts ("fibers-libevent[ERROR]: ", scm_current_error_port ());
+      scm_puts ("max events fired on [", scm_current_error_port ());
+      scm_display (scm_from_int(fd), scm_current_error_port ());
+      scm_puts ("], ignoring.\n", scm_current_error_port ());
+      return;
+    }
+
+  struct event_data ev_data = { fd, what };
+  memcpy (data->events + rv, &ev_data, sizeof (struct event_data));
+
+  data->rv += 1;
+}
+
+static SCM
+scm_primitive_event_wake (SCM wakefd)
+#define FUNC_NAME "primitive-event-wake"
+{
+  int c_fd;
+  char zero = 0;
+
+  c_fd = scm_to_int (wakefd);
+
+  if (write (c_fd, &zero, 1) <= 0 && errno != EWOULDBLOCK && errno != EAGAIN)
+    SCM_SYSERROR;
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_create_event_base (SCM eventsv)
+#define FUNC_NAME "primitive-create-event-base"
+{
+  struct libevt_data *libevt;
+  struct wait_data *data;
+  struct event_base *base;
+
+  if ((base = event_base_new ()) == NULL)
+    scm_syserror ("couldn't instantiate event_base");
```

Comment:
Is 'instantiate' some technical terminology, or is this just "couldn't allocate event_base"?

Reply:
allocate sounds great. updated.